README.rst
|Astropy| |PyPI| |Travis_Build| |Issues| |RTD|

Cosmo Utils - Utilities for LSS Research
-----------------------------------------------

Repository with scripts used in my LSS research

License
-------

This project is Copyright (c) Victor Calderon and licensed under the terms of the BSD 3-Clause license. See the licenses folder for more information.

.. |Astropy| image:: http://img.shields.io/badge/powered%20by-AstroPy-orange.svg?style=flat
    :target: http://www.astropy.org
    :alt: Powered by Astropy Badge

.. |Release| image:: https://img.shields.io/github/release/vcalderon2009/cosmo_utils.svg
   :target: https://github.com/vcalderon2009/cosmo_utils/releases/latest
   :alt: Latest Release

.. |PyPI| image:: https://img.shields.io/pypi/v/cosmo-utils.svg
   :target: https://pypi.python.org/pypi/cosmo-utils
   :alt: PyPI Release

.. |Travis_Build| image:: https://travis-ci.org/vcalderon2009/cosmo_utils.svg?branch=master
   :target: https://travis-ci.org/vcalderon2009/cosmo_utils
   :alt: Build Status

.. |Issues| image:: https://img.shields.io/github/issues/vcalderon2009/cosmo_utils.svg
   :target: https://github.com/vcalderon2009/cosmo_utils/issues
   :alt: Open Issues

.. |RTD| image:: https://readthedocs.org/projects/cosmo-utils/badge/?version=latest
   :target: http://cosmo-utils.readthedocs.io/en/latest/?badge=latest
   :alt: Documentation Status



CHANGES.rst
0.1.65 (unreleased)
-----------------------

- Fixed functions in `cosmo_utils.utils.geometry`.

0.1.64 (2019-08-14)
-----------------------

- Fixed functions in `cosmo_utils.utils.geometry`.

0.1.63 (2019-08-14)
-----------------------

- Modified the functions from `cosmo_utils.utils.stats_funcs` and
  `cosmo_utils.utils.geometry`.

0.1.62 (2019-03-06)
-----------------------

- Added new function `cosmo_utils.mock_catalogues.shmr_funcs.Behroozi2010Relation`
   and `cosmo_utils.mock_catalogues.shmr_funcs.Moster2010Relation`

0.1.61 (2019-03-05)
-----------------------

- Fixed bug in `~cosmo_utils.utils.stats_funcs.sigma_calcs` in how it returns
  percentiles.

0.1.60 (2019-02-27)
-----------------------

- Fixed bug in `~cosmo_utils.utils.stats_funcs.Stats_one_arr` again, and
  added option for returning a `dict` with the output variables

0.1.59 (2019-02-27)
-----------------------

- Changes to the function `~cosmo_utils.utils.stats_funcs.Stats_one_arr`.
  Fixed issue with `bin_statval` argument within the function.

0.1.58 (2019-02-11)
-----------------------

- Changes to `~cosmo_utils.utils.work_paths.get_output_path`. Now all the
  catalogues are under the `external` directory.

0.1.57 (2019-02-11)
-----------------------

- Changes to `~cosmo_utils.utils.work_paths.get_output_path`. Now all the
  catalogues are under the `external` directory.

0.1.56 (2019-02-06)
-----------------------

- Removed `pairwise_counting` module (Cython version) from the main package

0.1.55 (2019-01-16)
-----------------------

- Modified the function `~cosmo_utils.ml.ml_utils.train_test_dataset` to accept
  `pandas.DataFrame` input variables.

0.1.54 (2019-01-13)
-----------------------

- Added new feature ``sigma_clf_c`` which corresponds to the scatter in
  ``log(L)`` for central galaxies in the `conditional luminosity function`.

0.1.53 (2018-12-10)
-----------------------

- Fixed bug in :py:func:`~cosmo_utils.mock_catalogues.abundance_matching.abundance_matching_f` function.

0.1.52 (2018-12-03)
-----------------------

- Fixed bug in :py:func:`~cosmo_utils.utils.web_utils.url_files_download` again.

0.1.51 (2018-12-03)
-----------------------

- Fixed bug in :py:func:`~cosmo_utils.utils.web_utils.url_files_download`

0.1.50 (2018-12-03)
-----------------------

- Added :py:func:`~cosmo_utils.utils.web_utils.url_file_list` and
  :py:func:`~cosmo_utils.utils.web_utils.url_files_download` to utilities.

0.1.48 (2018-11-24)
-----------------------

- Fixed bugs

0.1.47 (2018-11-24)
-----------------------

- Fixed bug in :py:func:`~cosmo_utils.utils.stats_funcs.Stats_one_arr`.

0.1.46 (2018-10-04)
-----------------------

- Updated function :py:class:`~cosmo_utils.mock_catalogues.hod_funcs.HOD` to accomodate for better use.

0.1.45 (2018-10-03)
-----------------------

- Added new class for determining average number of galaxies as function
  of logarithmic halo mass :py:class:`~cosmo_utils.mock_catalogues.hod_funcs.HOD`

0.1.44 (2018-09-19)
-----------------------

- New function :py:func:`~cosmo_utils.utils.gen_utils.array_insert`.
- Updated function :py:func:`~cosmo_utils.mock_catalogues.shmr_funcs.Behroozi_relation`
  and fixed bug.

0.1.43 (2018-06-02)
-----------------------

- Fixed another bug in
  :py:func:`~cosmo_utils.ml.ml_utils.scoring_methods` with the `None` type.

0.1.42 (2018-06-02)
-----------------------

- Fixed bug with `feat_arr` in 
  :py:func:`~cosmo_utils.ml.ml_utils.scoring_methods`

0.1.41 (2018-05-31)
-----------------------

- Added module :py:module:`~cosmo_utils.utils.gen_utils`
- Expanded functionality in :py:func:`~cosmo_utils.ml.ml_utils.data_preprocessing`
  and :py:func:`~cosmo_utils.ml.ml_utils.train_test_dataset`.

0.1.40 (2018-05-30)
-----------------------

- Fixed bugs in :py:func:`~cosmo_utils.ml.ml_utils.train_test_dataset`

0.1.39 (2018-05-30)
-----------------------

- Used Flake8 Lint to go over the style of the code, and fixed bugs along the way

0.1.38 (2018-05-30)
-----------------------

- Fixed bug found in :py:func:`~cosmo_utils.utils.file_readers.pandas_df_to_hdf5_file`

0.1.37 (2018-05-29)
-----------------------

- Changed range for `hod_n_valid` in :py:func:`~cosmo_utils.mock_catalogues.catls_utils.catl_sdss_merge`
  , :py:func:`~cosmo_utils.mock_catalogues.catls_utils.catl_sdss_dir` and 
  , :py:func:`~cosmo_utils.mock_catalogues.catls_utils.extract_catls`.

0.1.36 (2018-05-27)
-----------------------

- A change in the path in 
  :py:func:`~cosmo_utils.utils.work_paths.get_sdss_catl_dir` and
  :py:func:`~cosmo_utils.utils.work_paths.get_output_path`

0.1.35 (2018-05-27)
-----------------------

- Fixed issue with path in 
  :py:func:`~cosmo_utils.mock_catalogues.catls_utils.catl_sdss_merge`

0.1.32 (2018-05-27)
-----------------------

- Modified paths in :py:func:`~cosmo_utils.utils.work_paths.cookiecutter_paths`
- Fixed path in :py:func:`~cosmo_utils.mock_catalogues.catls_utils.catl_sdss_merge`.

0.1.31 (2018-05-26)
-----------------------

- Added path and more for *velocity bias*

0.1.30 (2018-05-23)
-----------------------

- Minor bug in :py:func:`~cosmo_utils.utils.file_utils.mark_parametrize` fixed.

0.1.29 (2018-05-23)
-----------------------

- Added decorator to loop over different set of values (:py:func:`~cosmo_utils.utils.file_utils.mark_parametrize`).
- Fixed docstrings.

0.1.28 (2018-05-21)
-----------------------

- Fixed bug with :py:func:`~cosmo_utils.utils.file_utils.Path_Folder`

0.1.27 (2018-05-21)
-----------------------

- Modified the modules imported in :py:func:`~cosmo_utils.mock_catalogues.spherematch.spherematch`

0.1.26 (2018-05-17)
-----------------------

- Added some useful functions related to machine learning.
- Fixed bugs in testing.

0.1.25 (2018-05-17)
-----------------------

- Introduced `pairwise` counting again.
- Fixed bug in :py:func:`~cosmo_utils.mock_catalogues.shmr_funcs.Behroozi_relation`

0.1.24 (2018-05-17)
-----------------------

- Importing modules in a different way

cosmo_utils/ml/ml_utils.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

# Victor Calderon
# Created      : 2018-05-17
# Last Modified: 2018-05-31
from __future__ import absolute_import, division, print_function
__author__     = ['Victor Calderon']
__copyright__  = ["Copyright 2018 Victor Calderon"]
__email__      = ['victor.calderon@vanderbilt.edu']
__maintainer__ = ['Victor Calderon']
__all__        = ["data_preprocessing",
                  "train_test_dataset",
                  "scoring_methods"]

# Importing modules
import scipy

import numpy as np

from cosmo_utils.utils import file_utils as fd
from cosmo_utils.utils import gen_utils as gu
from cosmo_utils.custom_exceptions import LSSUtils_Error

# ML modules
import sklearn.metrics          as skmetrics
import sklearn.model_selection  as skms
import sklearn.preprocessing    as skpre

# Extra modules

## Functions

# Data preprocessing
def data_preprocessing(feat_arr, pre_opt='min_max', reshape=False):
    """
    Preprocess the data used, in order to clean and make the data more
    suitable for the machine learning algorithms

    Parameters
    -----------
    feat_arr : `numpy.ndarray`
        Array of feature values. This array is used for training a
        ML algorithm.

    pre_opt : {'min_max', 'standard', 'normalize', 'no'} `str`, optional
        Type of preprocessing to do on `feat_arr`.

        Options:
            - 'min_max' : Turns `feat_arr` to values between (0,1)
            - 'standard' : Uses `~sklearn.preprocessing.StandardScaler` method
            - 'normalize' : Uses the `~sklearn.preprocessing.Normalizer` method
            - 'no' : No preprocessing on `feat_arr`

    reshape : `bool`, optional
        If True, it reshapes `feat_arr` into a 1d array if its shapes is
        equal to (ncols, 1), where `ncols` is the number of columns.
        This variable is set to `False` by default.

    Returns
    -----------
    feat_arr_scaled : `numpy.ndarray`
        Rescaled version of `feat_arr` based on the choice of `pre_opt`.

    Notes
    -----------
    For more information on how to pre-process your data, see
    `http://scikit-learn.org/stable/modules/preprocessing.html`_.
    """
    file_msg = fd.Program_Msg(__file__)
    ## Checking input parameters
    # `feat_arr`
    feat_arr_type_valid = (list, np.ndarray)
    if not (isinstance(feat_arr, feat_arr_type_valid)):
        msg = '{0} `feat_arr` ({1}) is not a valid input type'.format(
            file_msg, type(feat_arr))
        raise LSSUtils_Error(msg)
    # `pre_opt`
    pre_opt_valid = ['min_max', 'standard', 'normalize', 'no']
    if not (pre_opt in pre_opt_valid):
        msg = '{0} `pre_opt` ({1}) is not a valid input'.format(
            file_msg, pre_opt)
        raise LSSUtils_Error(msg)
    ##
    ## Reshaping `feat_arr`
    if reshape:
        feat_arr = gu.reshape_arr_1d(feat_arr)
    ##
    ## Scaling `feat_arr`
    if (pre_opt == 'min_max'):
        # Scaler
        scaler = skpre.MinMaxScaler(feature_range=(0, 1))
        # Rescaling
        feat_arr_scaled = scaler.fit_transform(feat_arr)
    ## Standardize Data
    if pre_opt == 'standard':
        # Scaler
        scaler = skpre.StandardScaler().fit(feat_arr)
        # Rescaling
        feat_arr_scaled = scaler.transform(feat_arr)
    ## Normalize Data
    if pre_opt == 'normalize':
        # Scaler
        scaler = skpre.Normalizer().fit(feat_arr)
        # Rescaling
        feat_arr_scaled = scaler.transform(feat_arr)
    ## No Preprocessing
    if pre_opt == 'no':
        feat_arr_scaled = feat_arr

    return feat_arr_scaled

# Train-Test Data Split
def train_test_dataset(pred_arr, feat_arr, pre_opt='min_max',
    shuffle_opt=True, random_state=0, test_size=0.25, reshape=False):
    """
    Function to create the training and testing datasets for a given set
    of features array and predicted array.

    Parameters
    -----------
    pred_arr : `numpy.ndarray` or array-like, shape (n_samples, n_outcomes)
        Array consisting of the `predicted values`. The dimensions of
        `pred_arr` are `n_samples` by `n_outcomes`, where `n_samples` is the
        number of observations, and `n_outcomes` the number of predicted
        outcomes.

    feat_arr : `numpy.ndarray` or array-like, shape (n_samples, n_features)
        Array consisting of the `predicted values`. The dimensions of
        `feat_arr` are `n_samples` by `n_features`, where `n_samples`
        is the number of observations, and `n_features` the number of
        features used.

    pre_opt : {'min_max', 'standard', 'normalize', 'no'} `str`, optional
        Type of preprocessing to do on `feat_arr`.

        Options:
            - 'min_max' : Turns `feat_arr` to values between (0,1)
            - 'standard' : Uses `sklearn.preprocessing.StandardScaler` method
            - 'normalize' : Uses the `sklearn.preprocessing.Normalizer` method
            - 'no' : No preprocessing on `feat_arr`

    shuffle_opt : `bool`, optional
        If True, the data is shuffled before splitting into testing and
        training datasets. This variable is set to True by default.

    random_state : int, optional
        Random state number used for when splitting into training and
        testing datasets. If set, it will always have the same seed
        `random_state`. This variable is set to `0` by default.

    test_size : float, optional
        Percentage of the catalogue that represents the `test` size of
        the testing dataset. This variable must be between (0,1).
        This variable is set to `0.25` by default.

    reshape : `bool`, optional
        If True, it reshapes `feat_arr` into a 1d array if its shapes is
        equal to (ncols, 1), where `ncols` is the number of columns.
        This variable is set to `False` by default.

    Returns
    -----------
    train_dict : `dict`
        Dictionary containing the `training` data from the catalogue.

    test_dict : `dict`
        Dictionary containing the `testing` data from the catalogue.

    See also
    -----------
    data_preprocessing : Function to preprocess a dataset.
    """
    file_msg = fd.Program_Msg(__file__)
    ## Checking input parameters
    # `pred_arr`
    pred_arr_type_valid = (list, np.ndarray)
    if not (isinstance(pred_arr, pred_arr_type_valid)):
        msg = '{0} `pred_arr` ({1}) is not a valid input type'.format(
            file_msg, type(pred_arr))
        raise LSSUtils_Error(msg)
    # `feat_arr`
    feat_arr_type_valid = (list, np.ndarray)
    if not (isinstance(feat_arr, feat_arr_type_valid)):
        msg = '{0} `feat_arr` ({1}) is not a valid input type'.format(
            file_msg, type(feat_arr))
        raise LSSUtils_Error(msg)
    # `pre_opt`
    pre_opt_valid = ['min_max', 'standard', 'normalize', 'no']
    if not (pre_opt in pre_opt_valid):
        msg = '{0} `pre_opt` ({1}) is not a valid input'.format(
            file_msg, pre_opt)
        raise LSSUtils_Error(msg)
    # `shuffle_opt`
    shuffle_opt_type_valid = (bool)
    if not (isinstance(shuffle_opt, shuffle_opt_type_valid)):
        msg = '{0} `shuffle_opt` ({1}) is not a valid input type'.format(
            file_msg, type(shuffle_opt))
        raise LSSUtils_Error(msg)
    # `random_state`
    random_state_type_valid = (int)
    if not (isinstance(random_state, random_state_type_valid)):
        msg = '{0} `random_state` ({1}) is not a valid input'.format(
            file_msg, random_state)
        raise LSSUtils_Error(msg)
    # `test_size`
    if not ((test_size > 0) and (test_size < 1.)):
        msg = '{0} `test_size` ({1}) must be in range (0,1)'.format(
            file_msg, test_size)
        raise LSSUtils_Error(msg)
    ##
    ## Checking dimensions of `pred_arr` and `feat_arr`
    pred_arr = np.asarray(pred_arr)
    feat_arr = np.asarray(feat_arr)
    # Dimensions
    if reshape:
        pred_arr = gu.reshape_arr_1d(pred_arr)
        feat_arr = gu.reshape_arr_1d(feat_arr)
    # Shape
    if (len(pred_arr) != len(feat_arr)):
        msg  = '{0} The shape of `pred_arr` ({1}) and `feat_arr` ({2}) must '
        msg += 'have the same length'
        msg  = msg.format(file_msg, len(pred_arr), len(feat_arr))
        raise LSSUtils_Error(msg)
    ##
    ## Rescaling Dataset
    feat_arr_scaled = data_preprocessing(feat_arr, pre_opt=pre_opt,
        reshape=reshape)
    ##
    ## Splitting into `Training` and `Testing` datasets.
    # Scaled
    (   X_train, X_test,
        Y_train, Y_test) = skms.train_test_split(   feat_arr_scaled,
                                                    pred_arr,
                                                    test_size=test_size,
                                                    shuffle=shuffle_opt,
                                                    random_state=random_state)
    # Not-scaled
    (   X_train_ns, X_test_ns,
        Y_train_ns, Y_test_ns) = skms.train_test_split( feat_arr,
                                                        pred_arr,
                                                        test_size=test_size,
                                                        shuffle=shuffle_opt,
                                                        random_state=random_state)
    ##
    ## Assigning `training` and `testing` datasets to dictionaries
    train_dict = {  'X_train': X_train, 'Y_train': Y_train,
                    'X_train_ns': X_train_ns, 'Y_train_ns': Y_train_ns}
    test_dict  = {'X_test': X_test, 'Y_test': Y_test,
                    'X_test_ns': X_test_ns, 'Y_test_ns': Y_test_ns}

    return train_dict, test_dict

# Scoring methods
def scoring_methods(truth_arr, feat_arr=None, pred_arr=None, model=None,
    score_method='perc', threshold=0.1, perc=0.68):
    """
    Determines the overall score for given arrays, i.e. the `predicted`
    array and the `truth` array

    Parameters
    -----------
    truth_arr : `numpy.ndarray` or array-like, shape (n_samples, n_outcomes)
        Array consisting of the `true` values for the `n_samples`
        observations. The dimensions of `truth_arr` are
        `n_samples` by `n_outcomes`, where `n_samples` is the
        number of observations, and `n_outcomes` the number of predicted
        outcomes.

    feat_arr : `numpy.ndarray`, array-like, or `NoneType`, shape (n_samples, n_features)
        Array consisting of the `predicted values`. The dimensions of
        `feat_arr` are `n_samples` by `n_features`, where `n_samples`
        is the number of observations, and `n_features` the number of
        features used. This variable is set to `None` by default.

    pred_arr : `numpy.ndarray`, array-like, or `NoneType`, shape (n_samples, n_outcomes)
        Array of predicted values from `feat_arr`. If ``model == None``,
        this variable must be an array-like object. If ``model != None``,
        this variable will not be used, and will be calculated using
        the `model` object. This variable is set to `None` by default.

    model : scikit-learn model object or `NoneType`
        Model used to estimate the score if ``score_method == 'model_score'``
        This variable is set to `None` by default.

    score_method : {'perc', 'threshold', 'model_score', 'r2'} `str`, optional
        Type of scoring to use when determining how well an algorithm
        is performing.

        Options:
            - 'perc' : Use percentage and rank-ordering of the values
            - 'threshold' : Score based on diffs of `threshold` or less from true value.
            - 'model_score' : Out-of-the-box metod from `sklearn` to determine success.
            - 'r2': R-squared statistic for error calcuation.

    threshold : float, optional
        Value to use when calculating the error within `threshold` value
        from the truth. This variable is set to `0.1` by default.

    perc : float, optional
        Value used when determining score within some `perc` percentile
        value form [0,1]. This variable is set to `0.68` by default.

    Returns
    -----------
    method_score : float
        Overall score from `pred_arr` to predict `truth_arr`.

    Notes
    -----------
    For more information on how to pre-process your data, see
    `http://scikit-learn.org/stable/modules/model_evaluation.html`_.
    """
    file_msg = fd.Program_Msg(__file__)
    ## Checking input parameters
    # `feat_arr`
    feat_arr_type_valid = (list, np.ndarray, type(None))
    if not (isinstance(feat_arr, feat_arr_type_valid)):
        msg = '{0} `feat_arr` ({1}) is not a valid input type'.format(
            file_msg, type(feat_arr))
        raise LSSUtils_Error(msg)
    # `truth_arr`
    truth_arr_type_valid = (list, np.ndarray)
    if not (isinstance(truth_arr, truth_arr_type_valid)):
        msg = '{0} `truth_arr` ({1}) is not a valid input type'.format(
            file_msg, type(truth_arr))
        raise LSSUtils_Error(msg)
    # `score_method` - Type
    score_method_type_valid = (str)
    if not (isinstance(score_method, score_method_type_valid)):
        msg = '{0} `score_method` ({1}) is not a valid input type'.format(
            file_msg, type(score_method))
        raise LSSUtils_Error(msg)
    # `score_method` - Value
    score_method_valid = ['perc', 'threshold', 'model_score', 'r2']
    if not (score_method in score_method_valid):
        msg = '{0} `score_method` ({1}) is not a valid input!'.format(
            file_msg, score_method)
        raise LSSUtils_Error(score_method)
    # `threshold` - Type
    threshold_valid = (float, int)
    if not (isinstance(threshold, threshold_valid)):
        msg = '{0} `threshold` ({1}) is not a valid input type'.format(
            file_msg, type(threshold))
        raise LSSUtils_Error(msg)
    # `threshold` - Value
    if not (threshold >= 0.):
        msg = '{0} `threshold` ({1}) must be larger than 0!'.format(
            file_msg, threshold)
        raise LSSUtils_Error(msg)
    ##
    ## Checking for `model`, `pred_arr` and `feat_arr`
    # If both are none
    if ((model is None) and (pred_arr is None)):
        msg  = '{0} `model` and `pred_arr` cannot both be `None`. '
        msg += 'Only one can be `None`'
        msg  = msg.format(file_msg)
        raise LSSUtils_Error(msg)
    # If `feat_arr` and `pred_arr` are `None`
    if ((feat_arr is None) and (pred_arr is None)):
        msg = '{0} `feat_arr` and `pred_arr` cannot both be `None`'.format(
            file_msg)
        raise TypeError(msg)
    # `pred_arr` - Type
    # If both are `None`
    pred_arr_valid = ((list, np.ndarray))
    if (model is None):
        if not (isinstance(pred_arr, pred_arr_valid)):
            msg = '{0} `pred_arr` ({1}) is not a valid input type!'.format(
                file_msg, type(pred_arr))
            raise LSSUtils_Error(msg)
    ##
    ## Choosing scoring method
    # Percentile method
    if (score_method == 'perc'):
        # Checking for `pred_arr`
        if (pred_arr is None):
            pred_arr = model.predict(feat_arr)
        # Checking for `model`
        if (model is None):
            pred_arr = np.asarray(pred_arr)
        # Error calcualtion
        pred_err     = np.abs(pred_arr - truth_arr)
        method_score = scipy.stats.scoreatpercentile(pred_err, 100. * perc)
    # Threshold method
    if (score_method == 'threshold'):
        # Checking for `pred_arr`
        if (pred_arr is None):
            pred_arr = model.predict(feat_arr)
        # Checking for `model`
        if (model is None):
            pred_arr = np.asarray(pred_arr)
        # Error calcualtion
        pred_err     = np.abs(pred_arr - truth_arr)
        pred_thresh  = len(pred_err[pred_err <= threshold])
        method_score = pred_thresh / len(pred_arr)
    # R-squared method
    if (score_method == 'r2'):
        # Checking for `pred_arr`
        if (pred_arr is None):
            pred_arr = model.predict(feat_arr)
        # Checking for `model`
        if (model is None):
            pred_arr = np.asarray(pred_arr)
        # Error calcualtion
        method_score = skmetrics.r2_score(truth_arr, pred_arr)
    # Model method
    if (score_method == 'model_score'):
        method_score = model.score(feat_arr, truth_arr)

    return method_score
